packages/compass-app-stores/src/stores/instance-store.ts

```diff
@@ -0,0 +1,669 @@
+      );
+      return;
+    }
+
+    const coll = await instance.getNamespace({
+      database,
+      collection,
+      dataService,
+    });
+
+    return await coll?.fetchMetadata({ dataService });
+  }
+
+  async function refreshNamespace({
+    ns,
+    database,
+  }: {
+    ns: string;
+    database: string;
+  }) {
+    const instance = getInstance();
+    if (!instance.databases.get(database)) {
+      await instance.fetchDatabases({ dataService, force: true });
+    }
+    const db = instance.databases.get(database);
+    if (!db?.collections.get(ns)) {
+      await db?.fetchCollections({ dataService, force: true });
+    }
+    await refreshNamespaceStats(ns);
+  }
+
+  async function refreshNamespaceStats(ns: string) {
+    const instance = getInstance();
+    const { database } = toNS(ns);
+    const db = instance.databases.get(database);
+    const coll = db?.collections.get(ns);
+    // We don't care if this fails
+    await Promise.allSettled([
+      db?.fetch({ dataService, force: true }),
+      coll?.fetch({ dataService, force: true }),
+    ]);
+  }
+
+  const connectionString = dataService.getConnectionString();
+  const firstHost = connectionString.hosts[0] || '';
+  const [hostname, port] = firstHost.split(':');
+
+  const initialInstanceProps: Partial<MongoDBInstanceProps> = {
+    _id: firstHost,
+    hostname: hostname,
+    port: port ? +port : undefined,
+    topologyDescription: getTopologyDescription(
+      dataService.getLastSeenTopology()
+    ),
+  };
+  const instance = new MongoDBInstance(
+    initialInstanceProps as MongoDBInstanceProps
+  );
+  if ((globalThis as any).hadronApp) {
+    (globalThis as any).hadronApp.instance = instance;
+  }
+
+  debug('instance-created');
+  appRegistry.emit('instance-created', { instance });
+
+  store.dispatch(changeDataService(dataService));
+  store.dispatch(changeInstance(instance));
+
+  void refreshInstance({
+    fetchDatabases: true,
+    fetchDbStats: true,
+  });
+
+  function onTopologyDescriptionChanged({
+    newDescription,
+  }: {
+    newDescription: ReturnType<DataService['getLastSeenTopology']>;
+  }) {
+    (instance as any).set({
+      topologyDescription: getTopologyDescription(newDescription),
+    });
+  }
+  dataService.on('topologyDescriptionChanged', onTopologyDescriptionChanged);
+  cleanup.push(() =>
+    dataService.off?.(
+      'topologyDescriptionChanged',
+      onTopologyDescriptionChanged
+    )
+  );
+
+  cleanup.push(() => {
+    const hadronApp = (globalThis as any).hadronApp;
+    if (hadronApp?.instance === instance) {
+      hadronApp.instance?.removeAllListeners();
+      hadronApp.instance = null;
+    }
+    appRegistry.emit('instance-destroyed', { instance: null });
+    store.dispatch(reset());
+  });
+
+  const onSelectDatabase = voidify(fetchDatabaseDetails);
+  appRegistry.on('select-database', onSelectDatabase);
+  cleanup.push(() =>
+    appRegistry.removeListener('select-database', onSelectDatabase)
+  );
```

Comment:
Just a though for future: based on a few plugins we refactored already I wonder if we should change the app registry emitter methods interface to return unsubscribe method, we seem to be doing this literally every time we subscribe to something 🤔

Reply:
Fwiw, feel free to take a look at a45fdb679542fe88f42124bfc7ba5f6a9443bf14, that simplifies things a bit for this specific activate() function, but yeah, I’m happy to take it a bit further to make it more ergonomic

Reply:
Like the idea of this method! This plugin probably does most of the subscribtions by it's nature, so maybe let's do a few more plugins and see if we want to create something shared for these types of subscriptions